_source/_posts/2024-04-30-express-universal-logout.md

```diff
@@ -0,0 +1,674 @@
+
+### End a user's session
+
+In Express, we'll need to access a user's session in the session store from the express-session library. Create a new file called sessionStore.ts under the `apps/api/src` folder and create a variable called store. A separate file to reference the `store` variable in multiple files. 
+
+```ts
+import {MemoryStore} from 'express-session';
+export const store = new MemoryStore()
+```
+
+For now, import `store` from the `session/Store` file at the top of the `apps/api/src/main.ts`. 
+
+```ts
+import { store } from './sessionsStore';
+```
+
+The top of the `apps/api/src/main.ts` file will now look like this:
+
+```ts
+import express from 'express';
+import { PrismaClient, Todo, User } from '@prisma/client';
+import passportLocal from 'passport-local';
+import passportOIDC from 'passport-openidconnect';
+import passport from 'passport';
+import session from 'express-session';
+import { universalLogoutRoute } from './universalLogout';
+import morgan from 'morgan';
+
+// Bearer Auth - Universal Logout
+import passportBearer from 'passport-http-bearer';
+import { store } from './sessionsStore';
+```
+
+Now add `store` within your session configuration:
+
+```ts
+app.use(session({
+ resave: false,
+ saveUninitialized: false,
+ secret: 'top secret',
+ cookie: {
+   http: false,
+   sameSite: 'lax'
+ },
+ store
+}));
+```
+
+Notice how the user session is configured in the `apps/api/src/main.ts` file via the Passport.js library's serialization and deserialization functions. We can instruct the Passport.js library to reference a user's session via user.id; this will come in handy soon. If you are curious, you can read more about how this works in the [Passport.js documentation](https://www.passportjs.org/concepts/authentication/sessions/). 
+
+```ts
+passport.serializeUser( async (user: IUser, done) => {
+ done(null, user.id);
+});
+
+passport.deserializeUser( async (id: number, done) => {
+ const user: User = await prisma.user.findUnique({
+   where: {
+    id
+   }
+ });
+
+ done(null, user);
+});
+```
+
+Next, import the variable store at the top of the `apps/api/src/universalLogout.ts` file. 
+
+```ts
+import { Router } from 'express';
+export const universalLogoutRoute = Router();
+import { PrismaClient } from '@prisma/client';
+import { store } from './sessionsStore';
+```
+
+So far, we know the user's email from the request to the UL endpoint and have associated the user with an org via the provided API key. Now that we have the user object, we can associate the user's user ID with the session object. Add the following code to find and terminate a user's session.
+
+```ts
+// End user session
+  const storedSession = store.sessions;
+  const userId = user.id;
+  const sids = [];
+  Object.keys(storedSession).forEach((key) => {
+    const sess = JSON.parse(storedSession[key]);
+    if (sess.passport.user === userId) {
+      sids.push(key);
+    }
+  });
+
+  sids.map((sid) => store.destroy(sid));
```

Comment:
IMO I think this whole section's code can be optimized, but at the very least, we should use a `forEach` or `for of` to destroy. map implies you're mapping contents of the `sids` array to another type or format. that's not what's going on.

Reply:
Thank you, will fix this on my next iteration of improvements.